sno/structs.py

```diff
@@ -72,9 +72,39 @@ def resolve_refish(repo, refish):
         return CommitWithReference(repo.resolve_refish(refish))
 
 
-# When merging two commits, we have three versions of lots of things - these three versions
-# are the common ancestor, and the two versions to be merged, "ours" and "theirs".
-AncestorOursTheirs = namedtuple("AncestorOursTheirs", ("ancestor", "ours", "theirs"))
+# pygit2 always has this order - we use it too for consistency,
+# and so we can meaningfully zip() our tuples with theirs
+_ANCESTOR_OURS_THEIRS_ORDER = ("ancestor", "ours", "theirs")
 
-AncestorOursTheirs.names = AncestorOursTheirs._fields
-AncestorOursTheirs.chars = tuple(n[0] for n in AncestorOursTheirs.names)
+
+class AncestorOursTheirs(namedtuple("AncestorOursTheirs", _ANCESTOR_OURS_THEIRS_ORDER)):
+    """
+    When merging two commits, we can end up with three versions of lots of things -
+    commits, repository-structures, datasets, features, primary keys.
+    The 3 versions  are the common ancestor, and 2 versions to be merged, "ours" and "theirs".
+    Like pygit2, we keep the 3 versions always in the same order - ancestor, ours, theirs.
+    """
+
+    NAMES = _ANCESTOR_OURS_THEIRS_ORDER
+    CHARS = tuple(n[0] for n in NAMES)
+
+    @staticmethod
+    def partial(ancestor=None, ours=None, theirs=None, *args):
```

Comment:
not sure why this method accepts `*args` when it just asserts there aren't any

Reply:
This method is supposed to be called with only kwargs - args have the wrong semantics - but without this assert, partial(), partial(1), partial(1, 2) would all be valid, but I have decided that is not how this method is used and so those calls are errors. (Maybe a similar idea to your comment earlier about how I had written an overly accepting method)

Reply:
I don't think this stops it being called with args, because args fill up the earlier arguments first even if they're named:

```python
>>>
... def partial(ancestor=None, ours=None, theirs=None, *args):
...     print(ancestor, ours, theirs, args)
...
...
>>> partial()
None None None ()
>>> partial(1)
1 None None ()

```

However, you _can_ specify keyword-only arguments, like:

```python
>>> def partial(*, ancestor=None, ours=None, theirs=None):
...     print(ancestor, ours, theirs)

>>> partial(1)
Traceback (most recent call last):
  File "<ipython-input-14-24db84d21a38>", line 1, in <module>
    partial(1)
TypeError: partial() takes 0 positional arguments but 1 was given

# doesn't help you with zero-args case though:
>>> def partial(*, ancestor=None, ours=None, theirs=None):
...     print(ancestor, ours, theirs)
>>> partial()
None None None
```